WTF::removeQueryParameters should not reencode and reparse the query string if nothing was removed

whsieh opened the pull request:
#### e9fd9cc2e7f15f9c93c833bfcb443d28da33276e
<pre>
WTF::removeQueryParameters should not reencode and reparse the query string if nothing was removed
<a href="https://bugs.webkit.org/show_bug.cgi?id=247131">https://bugs.webkit.org/show_bug.cgi?id=247131</a>

Reviewed by Chris Dumez.

Add a minor optimization to the `WTF::removeQueryParameters` helper function. In the case where no
query parameters were removed, we should leave that part of the query string alone, instead of
forcing both the key and value to become percent-encoded. Additionally, in the case where no query
parameters were removed altogether, we can skip setting (and re-parsing) the query string
altogether, and instead simply return without modifying the given URL.

* Source/WTF/wtf/URL.cpp:
(WTF::removeQueryParameters):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::parseURLEncodedForm):
(WTF::URLParser::parseQueryNameAndValue):
* Source/WTF/wtf/URLParser.h:
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:

Augment an existing API test to exercise this change.

(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/256078@main">https://commits.webkit.org/256078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5ebdc5a18ebf3cf36c93209c4c635f6778532f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3735 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/27462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/104207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98564 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3794 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/31930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/100176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100236 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80947 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/86886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/27462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/85755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38332 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/27462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80940 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/27462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/27815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4197 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40094 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83609 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42055 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/27462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18890 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -2883,21 +2883,28 @@ auto URLParser::parseURLEncodedForm(StringView input) -> URLEncodedForm
+        auto name = formURLDecode(makeStringByReplacingAll(bytes.left(equalIndex), '+', 0x20));
+        auto value = formURLDecode(makeStringByReplacingAll(bytes.substring(equalIndex + 1), '+', 0x20));
+        if (name && value)
+            return { { name.value(), value.value() } };
```
`return { { WTFMove(*name), WTFMove(*value) } };`

Would be a tad more efficient.

whsieh replied:
Fixed!

cdumez commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -2883,21 +2883,28 @@ auto URLParser::parseURLEncodedForm(StringView input) -> URLEncodedForm
+    if (equalIndex == notFound) {
+        auto name = formURLDecode(makeStringByReplacingAll(bytes, '+', 0x20));
+        if (name)
+            return { { name.value(), emptyString() } };
```
WTFMove(*name)

cdumez commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -2883,21 +2883,28 @@ auto URLParser::parseURLEncodedForm(StringView input) -> URLEncodedForm
+        if (name)
+            return { { name.value(), emptyString() } };
+    } else {
+        auto name = formURLDecode(makeStringByReplacingAll(bytes.left(equalIndex), '+', 0x20));
```
Maybe we can use ' ' instead of 0x20? It feels more readable.

whsieh replied:
Sounds good! Changed to `' '` here and below.

cdumez commented on Source/WTF/wtf/URLParser.cpp:
```diff
@@ -2883,21 +2883,28 @@ auto URLParser::parseURLEncodedForm(StringView input) -> URLEncodedForm
+            return { { name.value(), emptyString() } };
+    } else {
+        auto name = formURLDecode(makeStringByReplacingAll(bytes.left(equalIndex), '+', 0x20));
+        auto value = formURLDecode(makeStringByReplacingAll(bytes.substring(equalIndex + 1), '+', 0x20));
```
ditto.

cdumez reviewed (approved):
LGTM

whsieh committed:
WTF::removeQueryParameters should not reencode and reparse the query string if nothing was removed
https://bugs.webkit.org/show_bug.cgi?id=247131

Reviewed by Chris Dumez.

Add a minor optimization to the `WTF::removeQueryParameters` helper function. In the case where no
query parameters were removed, we should leave that part of the query string alone, instead of
forcing both the key and value to become percent-encoded. Additionally, in the case where no query
parameters were removed altogether, we can skip setting (and re-parsing) the query string
altogether, and instead simply return without modifying the given URL.

* Source/WTF/wtf/URL.cpp:
(WTF::removeQueryParameters):
* Source/WTF/wtf/URLParser.cpp:
(WTF::URLParser::parseURLEncodedForm):
(WTF::URLParser::parseQueryNameAndValue):
* Source/WTF/wtf/URLParser.h:
* Tools/TestWebKitAPI/Tests/WTF/URL.cpp:

Augment an existing API test to exercise this change.

(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/256078@main
